File: apps/api/src/data_intelligence_hub/schemas/workflow_planner.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from enum import StrEnum
from typing import Literal, Self
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, JsonValue, model_validator

from data_intelligence_hub.schemas.capability_catalog import (
    CapabilityOperation,
    CapabilityStatus,
    PlatformId,
    ResourceType,
)


class WorkflowPlannerContract(BaseModel):
    model_config = ConfigDict(extra="forbid")


class FlowMode(StrEnum):
    PERIODIC_MONITORING = "periodic_monitoring"
    BATCH_RESEARCH = "batch_research"


class MonitoringScopeType(StrEnum):
    BRAND = "brand"
    CATEGORY = "category"
    COMPETITOR = "competitor"
    TOPIC = "topic"
    CAMPAIGN = "campaign"


class MatchMode(StrEnum):
    EXACT = "exact"
    PHRASE = "phrase"
    SEMANTIC = "semantic"
    HYBRID = "hybrid"


class PolicyProfile(StrEnum):
    MARKET_MONITORING_BALANCED = "market_monitoring_balanced"


class AuthReadiness(StrEnum):
    NOT_REQUIRED = "not_required"
    READY = "ready"
    MISSING = "missing"
    NOT_CHECKED = "not_checked"


class PlanningStatus(StrEnum):
    RESOLVED = "resolved"
    PARTIALLY_RESOLVED = "partially_resolved"
    HELD = "held"


class RoutePlanStatus(StrEnum):
    RESOLVED = "resolved"
    PARTIAL = "partial"
    HELD = "held"


class WorkflowStepPlanningStatus(StrEnum):
    PLANNED = "planned"
    PARTIAL = "partial"
    HELD = "held"
    NOT_APPLICABLE = "not_applicable"


class BudgetStatus(StrEnum):
    WITHIN_CEILING = "within_ceiling"
    EXCEEDED = "exceeded"
    UNKNOWN = "unknown"
    NOT_APPLICABLE = "not_applicable"


class ScheduleIntent(WorkflowPlannerContract):
    cadence: Literal["hourly", "daily", "weekly"]
    timezone: str = Field(min_length=1, max_length=100)


class DeliveryIntent(WorkflowPlannerContract):
    outputs: list[Literal["dataset", "alert", "brief"]] = Field(min_length=1)


class BudgetCeiling(WorkflowPlannerContract):
    amount: Decimal = Field(ge=0)
    currency: Literal["USD"] = "USD"


class RateLimitIntent(WorkflowPlannerContract):
    max_requests: int = Field(ge=1)
    period_seconds: int = Field(ge=1)


class RetentionIntent(WorkflowPlannerContract):
    days: int = Field(ge=1, le=3650)


class MonitoringScopeDraft(WorkflowPlannerContract):
    scope_ref: str = Field(min_length=1, max_length=100)
    scope_type: MonitoringScopeType
    canonical_term: str | None = Field(default=None, max_length=500)
    aliases: list[str] = Field(default_factory=list, max_length=50)
    include_terms: list[str] = Field(default_factory=list, max_length=50)
    exclude_terms: list[str] = Field(default_factory=list, max_length=50)
    official_accounts: list[str] = Field(default_factory=list, max_length=50)
    seed_urls: list[str] = Field(default_factory=list, max_length=100)
    languages: list[str] = Field(default_factory=list)
    regions: list[str] = Field(default_factory=list)
    platforms: list[PlatformId] = Field(default_factory=list)
    match_mode: MatchMode | None = None

    @model_validator(mode="after")
    def validate_scope_content(self) -> Self:
        if self.scope_type in {
            MonitoringScopeType.BRAND,
            MonitoringScopeType.CATEGORY,
            MonitoringScopeType.COMPETITOR,
        } and not (self.canonical_term and self.canonical_term.strip()):
            raise ValueError("canonical_term_required")
        if self.scope_type in {
            MonitoringScopeType.TOPIC,
            MonitoringScopeType.CAMPAIGN,
        } and not any(
            (
                self.canonical_term and self.canonical_term.strip(),
                self.aliases,
                self.include_terms,
                self.official_accounts,
                self.seed_urls,
            )
        ):
            raise ValueError("scope_input_required")
        return self


class PlanningInput(WorkflowPlannerContract):
    flow_mode: FlowMode
    scopes: list[MonitoringScopeDraft] = Field(min_length=1, max_length=20)
    default_languages: list[str] = Field(default_factory=list)
    default_regions: list[str] = Field(default_factory=list)
    default_platforms: list[PlatformId] = Field(default_factory=list)
    schedule_intent: ScheduleIntent | None = None
    delivery_intent: DeliveryIntent | None = None
    policy_profile: PolicyProfile = PolicyProfile.MARKET_MONITORING_BALANCED
    purpose: Literal["brand_monitoring", "market_research", "competitive_research"]
    required_fields: list[str] = Field(default_factory=list)
    optional_fields: list[str] = Field(default_factory=list)
    budget_ceiling: BudgetCeiling | None = None
    rate_limit_intent: RateLimitIntent | None = None
    retention_intent: RetentionIntent | None = None
    allow_partial_degradation: bool = False

    @model_validator(mode="after")
    def validate_flow_contract(self) -> Self:
        refs = [scope.scope_ref for scope in self.scopes]
        if len(refs) != len(set(refs)):
            raise ValueError("duplicate_scope_ref")
        if sum(len(scope.seed_urls) for scope in self.scopes) > 100:
            raise ValueError("seed_url_limit_exceeded")
        if self.flow_mode is FlowMode.PERIODIC_MONITORING:
            if self.schedule_intent is None:
                raise ValueError("periodic_schedule_required")
            if any(
                not (scope.platforms or self.default_platforms or scope.seed_urls)
                for scope in self.scopes
            ):
                raise ValueError("periodic_platform_or_seed_url_required")
        else:
            if "schedule_intent" in self.model_fields_set:
                raise ValueError("batch_schedule_not_allowed")
            if not any(
                (
                    scope.canonical_term and scope.canonical_term.strip(),
                    scope.aliases,
                    scope.include_terms,
                    scope.official_accounts,
                    scope.seed_urls,
                )
                for scope in self.scopes
            ):
                raise ValueError("batch_input_required")
            if any(
                any(
                    (
                        scope.canonical_term and scope.canonical_term.strip(),
                        scope.aliases,
                        scope.include_terms,
                        scope.official_accounts,
                    )
                )
                and not (scope.platforms or self.default_platforms)
                for scope in self.scopes
            ):
                raise ValueError("batch_query_platform_required")
        return self


class NormalizedMonitoringScope(WorkflowPlannerContract):
    scope_key: str
    source_scope_refs: list[str]
    scope_type: MonitoringScopeType
    canonical_term: str | None
    aliases: list[str]
    include_terms: list[str]
    exclude_terms: list[str]
    official_accounts: list[str]
    seed_urls: list[str]
    effective_languages: list[str]
    effective_regions: list[str]
    effective_platforms: list[PlatformId]
    match_mode: MatchMode


class NormalizedPlanningInput(WorkflowPlannerContract):
    flow_mode: FlowMode
    scopes: list[NormalizedMonitoringScope]
    schedule_intent: ScheduleIntent | None
    delivery_intent: DeliveryIntent | None
    policy_profile: PolicyProfile
    purpose: Literal["brand_monitoring", "market_research", "competitive_research"]
    required_fields: list[str]
    optional_fields: list[str]
    budget_ceiling: BudgetCeiling | None
    rate_limit_intent: RateLimitIntent | None
    retention_intent: RetentionIntent | None
    allow_partial_degradation: bool


class ScopeRefMapping(WorkflowPlannerContract):
    scope_ref: str
    scope_key: str


class DecisionReason(WorkflowPlannerContract):
    code: str
    reason: str


class QueryTerm(WorkflowPlannerContract):
    term: str
    normalized_term: str
    scope_ref: str
    scope_key: str
    origin: Literal[
        "canonical",
        "alias",
        "include",
        "official_account",
        "seed_url",
        "fixture_candidate_expansion",
    ]
    status: Literal["active", "candidate", "rejected"]
    reason: str | None
    source: str
    score: float | None
    conflict_codes: list[str]


class SemanticQueryTerm(WorkflowPlannerContract):
    term: str
    normalized_term: str
    scope_key: str
    origin: Literal[
        "canonical",
        "alias",
        "include",
        "official_account",
        "seed_url",
        "fixture_candidate_expansion",
    ]
    status: Literal["active", "candidate", "rejected"]
    reason: str | None
    source: str
    score: float | None
    conflict_codes: list[str]


class CompiledPlatformQuery(WorkflowPlannerContract):
    platform: PlatformId
    scope_keys: list[str]
    source_scope_refs: list[str]
    resource_type: ResourceType
    operation: CapabilityOperation
    query_version: str
    normalized_expression: str
    include_terms: list[str]
    exclude_terms: list[str]
    account_filters: list[str]
    url_inputs: list[str]
    limitations: list[str]


class SemanticCompiledPlatformQuery(WorkflowPlannerContract):
    platform: PlatformId
    scope_keys: list[str]
    resource_type: ResourceType
    operation: CapabilityOperation
    query_version: str
    normalized_expression: str
    include_terms: list[str]
    exclude_terms: list[str]
    account_filters: list[str]
    url_inputs: list[str]
    limitations: list[str]


class QueryCompilerFailure(WorkflowPlannerContract):
    platform: PlatformId
    scope_keys: list[str]
    code: Literal["compiler_missing"] = "compiler_missing"
    reason: str


class RouteRequirement(WorkflowPlannerContract):
    requirement_ref: str
    scope_keys: list[str]
    step_refs: list[str]
    platform: PlatformId
    resource_type: ResourceType
    operation: CapabilityOperation
    purpose: Literal["brand_monitoring", "market_research", "competitive_research"]
    regions: list[str]
    required_fields: list[str]
    optional_fields: list[str]
    budget_ceiling: BudgetCeiling | None
    freshness_requirement: str | None
    rate_limit_requirement: RateLimitIntent | None
    retention_requirement: RetentionIntent | None
    allow_partial_degradation: bool
    precondition_failures: list[DecisionReason] = Field(default_factory=list)


class CapabilityReadinessSnapshot(WorkflowPlannerContract):
    implementation_id: str
    auth_readiness: AuthReadiness
    source: str
    credential_read_status: Literal["not_read"] = "not_read"


class ScoreBreakdown(WorkflowPlannerContract):
    raw_dimensions: dict[str, int]
    effective_dimensions: dict[str, int]
    weights: dict[str, int]
    weighted_score: int
    trace_codes: list[str]


class RouteCandidateDecision(WorkflowPlannerContract):
    assertion_id: str
    implementation_id: str
    capability_status: CapabilityStatus
    score_breakdown: ScoreBreakdown | None
    weighted_score: int | None
    route_eligible: bool
    readiness_status: AuthReadiness
    approval_required: bool
    approval_reasons: list[DecisionReason]
    missing_optional_fields: list[str]
    evidence_refs: list[str]


class ShadowRule(WorkflowPlannerContract):
    enabled: bool
    fallback_implementation_id: str | None
    sample_rate: float | None
    max_items: int | None
    reason: str
    execution_authorized: Literal[False] = False


class RoutePlanPreview(WorkflowPlannerContract):
    requirement_ref: str
    status: RoutePlanStatus
    primary_implementation: RouteCandidateDecision | None
    fallback_implementations: list[RouteCandidateDecision]
    shadow_rule: ShadowRule
    required_fields: list[str]
    optional_fields: list[str]
    missing_optional_fields: list[str]
    budget_status: BudgetStatus
    rate_limit_policy: RateLimitIntent | None
    retention_policy: RetentionIntent | None
    route_eligible: bool
    readiness_status: AuthReadiness | None
    approval_required: bool
    approval_reasons: list[DecisionReason]
    policy_gates: list[DecisionReason]
    score_breakdown: ScoreBreakdown | None
    exclusion_reasons: list[DecisionReason]
    degradation_rule: DecisionReason | None
    limitations: list[str]
    execution_authorized: Literal[False] = False


class StepDataContractField(WorkflowPlannerContract):
    name: str
    data_type: str
    cardinality: str
    required: bool
    source_step_ref: str | None
    description: str


class StepDataContract(WorkflowPlannerContract):
    schema_version: str
    fields: list[StepDataContractField]


class WorkflowStepPreview(WorkflowPlannerContract):
    step_ref: str
    template_key: str
    sequence: int = Field(ge=1)
    label: str
    execution_kind: Literal["planner_internal", "future_capability"]
    depends_on: list[str]
    platform: PlatformId | None
    scope_keys: list[str]
    resource_type: ResourceType | None
    operation: CapabilityOperation | None
    requirement_ref: str | None
    input_contract: StepDataContract
    output_contract: StepDataContract
    planning_status: WorkflowStepPlanningStatus
    limitations: list[str]


class DecisionTraceEntry(WorkflowPlannerContract):
    code: str
    reason: str
    scope_keys: list[str]
    requirement_ref: str | None
    details: dict[str, JsonValue]


class DecisionTrace(WorkflowPlannerContract):
    semantic_entries: list[DecisionTraceEntry]
    input_diagnostics: list[DecisionTraceEntry]


class CoverageSummary(WorkflowPlannerContract):
    total_requirements: int = Field(ge=0)
    resolved_requirements: int = Field(ge=0)
    partial_requirements: int = Field(ge=0)
    held_requirements: int = Field(ge=0)


class BudgetSummary(WorkflowPlannerContract):
    currency: Literal["USD"] = "USD"
    known_selected_unit_cost: Decimal | None
    unknown_count: int = Field(ge=0)
    budget_status: BudgetStatus


class AttributionContract(WorkflowPlannerContract):
    matched_scope_id: str
    matched_term: str
    match_reason: str
    query_version: str
    requirement_ref: str
    route_plan_ref: str


FINGERPRINT_FORBIDDEN_REFERENCE_FIELDS = frozenset(
    {
        "scope_ref",
        "source_scope_refs",
        "project_id",
        "generated_at",
        "request_id",
    }
)


def _reject_fingerprint_reference_fields(value: object) -> None:
    if isinstance(value, BaseModel):
        _reject_fingerprint_reference_fields(value.model_dump(mode="python"))
    elif isinstance(value, dict):
        for key, nested_value in value.items():
            if key in FINGERPRINT_FORBIDDEN_REFERENCE_FIELDS:
                raise ValueError(f"fingerprint_reference_field_forbidden:{key}")
            _reject_fingerprint_reference_fields(nested_value)
    elif isinstance(value, (list, tuple)):
        for item in value:
            _reject_fingerprint_reference_fields(item)


class WorkflowPlanFingerprintPayload(WorkflowPlannerContract):
    planner_contract_version: str
    fingerprint_input: dict[str, JsonValue]
    catalog_snapshot_id: str
    policy_version: str
    mode_template_version: str
    query_versions: dict[PlatformId, str]
    candidate_fixture_version: str
    semantic_query_terms: list[SemanticQueryTerm]
    semantic_steps: list[WorkflowStepPreview]
    semantic_compiled_queries: list[SemanticCompiledPlatformQuery]
    route_plans: list[RoutePlanPreview]
    coverage: CoverageSummary
    budget_summary: BudgetSummary
    limitations: list[str]
    semantic_decision_trace: list[DecisionTraceEntry]

    @model_validator(mode="after")
    def validate_reference_free(self) -> Self:
        _reject_fingerprint_reference_fields(self)
        return self


class WorkflowPlanPreview(WorkflowPlannerContract):
    schema_version: Literal["workflow_plan_preview.v1"]
    planner_contract_version: str
    project_id: UUID
    flow_mode: FlowMode
    planning_status: PlanningStatus
    normalized_input: NormalizedPlanningInput
    scope_ref_map: list[ScopeRefMapping]
    query_terms: list[QueryTerm]
    compiled_queries: list[CompiledPlatformQuery]
    steps: list[WorkflowStepPreview]
    route_requirements: list[RouteRequirement]
    route_plans: list[RoutePlanPreview]
    coverage: CoverageSummary
    budget_summary: BudgetSummary
    limitations: list[str]
    decision_trace: DecisionTrace
    attribution_contract: AttributionContract
    catalog_snapshot_id: str
    policy_version: str
    mode_template_version: str
    query_versions: dict[PlatformId, str]
    preview_fingerprint: str
    execution_authorized: Literal[False] = False
    provider_call: Literal[False] = False
    actor_run: Literal[False] = False
    browser_run: Literal[False] = False
    llm_call: Literal[False] = False
    workflow_run_created: Literal[False] = False
    database_write: Literal[False] = False
    generated_at: datetime
    request_id: str

File: apps/web/src/types/workflow-planner.ts
import type {
  CapabilityOperation,
  CapabilityPlatform,
  CapabilityResourceType,
  CapabilityStatus,
} from "@/types/capability";

export type WorkflowPlannerMode = "periodic_monitoring" | "batch_research";
export type MonitoringScopeType =
  | "brand"
  | "category"
  | "competitor"
  | "topic"
  | "campaign";
export type WorkflowPlannerMatchMode =
  | "exact"
  | "phrase"
  | "semantic"
  | "hybrid";
export type WorkflowPlannerPolicyProfile = "market_monitoring_balanced";
export type WorkflowPlannerPurpose =
  | "brand_monitoring"
  | "market_research"
  | "competitive_research";
export type WorkflowPlannerAuthReadiness =
  | "not_required"
  | "ready"
  | "missing"
  | "not_checked";
export type WorkflowPlanningStatus =
  | "resolved"
  | "partially_resolved"
  | "held";
export type RoutePlanStatus = "resolved" | "partial" | "held";
export type WorkflowStepPlanningStatus =
  | "planned"
  | "partial"
  | "held"
  | "not_applicable";
export type WorkflowBudgetStatus =
  | "within_ceiling"
  | "exceeded"
  | "unknown"
  | "not_applicable";

export type PlannerJsonValue =
  | string
  | number
  | boolean
  | null
  | PlannerJsonValue[]
  | { [key: string]: PlannerJsonValue };

export type ScheduleIntent = {
  cadence: "hourly" | "daily" | "weekly";
  timezone: string;
};

export type ScheduleIntentDto = ScheduleIntent;

export type DeliveryIntent = {
  outputs: Array<"dataset" | "alert" | "brief">;
};

export type DeliveryIntentDto = DeliveryIntent;

export type BudgetCeiling = {
  amount: string;
  currency: "USD";
};

export type BudgetCeilingDto = BudgetCeiling;

export type RateLimitIntent = {
  maxRequests: number;
  periodSeconds: number;
};

export type RateLimitIntentDto = {
  max_requests: number;
  period_seconds: number;
};

export type RetentionIntent = {
  days: number;
};

export type RetentionIntentDto = RetentionIntent;

export type MonitoringScopeDraft = {
  scopeRef: string;
  scopeType: MonitoringScopeType;
  canonicalTerm: string | null;
  aliases: string[];
  includeTerms: string[];
  excludeTerms: string[];
  officialAccounts: string[];
  seedUrls: string[];
  languages: string[];
  regions: string[];
  platforms: CapabilityPlatform[];
  matchMode: WorkflowPlannerMatchMode | null;
};

export type MonitoringScopeDraftDto = {
  scope_ref: string;
  scope_type: MonitoringScopeType;
  canonical_term: string | null;
  aliases: string[];
  include_terms: string[];
  exclude_terms: string[];
  official_accounts: string[];
  seed_urls: string[];
  languages: string[];
  regions: string[];
  platforms: CapabilityPlatform[];
  match_mode: WorkflowPlannerMatchMode | null;
};

type PlanningInputBase = {
  scopes: MonitoringScopeDraft[];
  defaultLanguages: string[];
  defaultRegions: string[];
  defaultPlatforms: CapabilityPlatform[];
  deliveryIntent: DeliveryIntent | null;
  policyProfile: WorkflowPlannerPolicyProfile;
  purpose: WorkflowPlannerPurpose;
  requiredFields: string[];
  optionalFields: string[];
  budgetCeiling: BudgetCeiling | null;
  rateLimitIntent: RateLimitIntent | null;
  retentionIntent: RetentionIntent | null;
  allowPartialDegradation: boolean;
};

export type PeriodicPlanningInput = PlanningInputBase & {
  flowMode: "periodic_monitoring";
  scheduleIntent: ScheduleIntent;
};

export type BatchPlanningInput = PlanningInputBase & {
  flowMode: "batch_research";
  scheduleIntent?: never;
};

export type PlanningInput = PeriodicPlanningInput | BatchPlanningInput;

type PlanningInputDtoBase = {
  scopes: MonitoringScopeDraftDto[];
  default_languages: string[];
  default_regions: string[];
  default_platforms: CapabilityPlatform[];
  delivery_intent: DeliveryIntentDto | null;
  policy_profile: WorkflowPlannerPolicyProfile;
  purpose: WorkflowPlannerPurpose;
  required_fields: string[];
  optional_fields: string[];
  budget_ceiling: BudgetCeilingDto | null;
  rate_limit_intent: RateLimitIntentDto | null;
  retention_intent: RetentionIntentDto | null;
  allow_partial_degradation: boolean;
};

export type PeriodicPlanningInputDto = PlanningInputDtoBase & {
  flow_mode: "periodic_monitoring";
  schedule_intent: ScheduleIntentDto;
};

export type BatchPlanningInputDto = PlanningInputDtoBase & {
  flow_mode: "batch_research";
  schedule_intent?: never;
};

export type PlanningInputDto =
  | PeriodicPlanningInputDto
  | BatchPlanningInputDto;

export type NormalizedMonitoringScopeDto = {
  scope_key: string;
  source_scope_refs: string[];
  scope_type: MonitoringScopeType;
  canonical_term: string | null;
  aliases: string[];
  include_terms: string[];
  exclude_terms: string[];
  official_accounts: string[];
  seed_urls: string[];
  effective_languages: string[];
  effective_regions: string[];
  effective_platforms: CapabilityPlatform[];
  match_mode: WorkflowPlannerMatchMode;
};

export type NormalizedMonitoringScope = {
  scopeKey: string;
  sourceScopeRefs: string[];
  scopeType: MonitoringScopeType;
  canonicalTerm: string | null;
  aliases: string[];
  includeTerms: string[];
  excludeTerms: string[];
  officialAccounts: string[];
  seedUrls: string[];
  effectiveLanguages: string[];
  effectiveRegions: string[];
  effectivePlatforms: CapabilityPlatform[];
  matchMode: WorkflowPlannerMatchMode;
};

export type NormalizedPlanningInputDto = {
  flow_mode: WorkflowPlannerMode;
  scopes: NormalizedMonitoringScopeDto[];
  schedule_intent: ScheduleIntentDto | null;
  delivery_intent: DeliveryIntentDto | null;
  policy_profile: WorkflowPlannerPolicyProfile;
  purpose: WorkflowPlannerPurpose;
  required_fields: string[];
  optional_fields: string[];
  budget_ceiling: BudgetCeilingDto | null;
  rate_limit_intent: RateLimitIntentDto | null;
  retention_intent: RetentionIntentDto | null;
  allow_partial_degradation: boolean;
};

export type NormalizedPlanningInput = {
  flowMode: WorkflowPlannerMode;
  scopes: NormalizedMonitoringScope[];
  scheduleIntent: ScheduleIntent | null;
  deliveryIntent: DeliveryIntent | null;
  policyProfile: WorkflowPlannerPolicyProfile;
  purpose: WorkflowPlannerPurpose;
  requiredFields: string[];
  optionalFields: string[];
  budgetCeiling: BudgetCeiling | null;
  rateLimitIntent: RateLimitIntent | null;
  retentionIntent: RetentionIntent | null;
  allowPartialDegradation: boolean;
};

export type ScopeRefMappingDto = {
  scope_ref: string;
  scope_key: string;
};

export type ScopeRefMapping = {
  scopeRef: string;
  scopeKey: string;
};

export type DecisionReason = {
  code: string;
  reason: string;
};

export type DecisionReasonDto = DecisionReason;

export type QueryTermDto = {
  term: string;
  normalized_term: string;
  scope_ref: string;
  scope_key: string;
  origin:
    | "canonical"
    | "alias"
    | "include"
    | "official_account"
    | "seed_url"
    | "fixture_candidate_expansion";
  status: "active" | "candidate" | "rejected";
  reason: string | null;
  source: string;
  score: number | null;
  conflict_codes: string[];
};

export type QueryTerm = {
  term: string;
  normalizedTerm: string;
  scopeRef: string;
  scopeKey: string;
  origin: QueryTermDto["origin"];
  status: QueryTermDto["status"];
  reason: string | null;
  source: string;
  score: number | null;
  conflictCodes: string[];
};

export type CompiledPlatformQueryDto = {
  platform: CapabilityPlatform;
  scope_keys: string[];
  source_scope_refs: string[];
  resource_type: CapabilityResourceType;
  operation: CapabilityOperation;
  query_version: string;
  normalized_expression: string;
  include_terms: string[];
  exclude_terms: string[];
  account_filters: string[];
  url_inputs: string[];
  limitations: string[];
};

export type CompiledPlatformQuery = {
  platform: CapabilityPlatform;
  scopeKeys: string[];
  sourceScopeRefs: string[];
  resourceType: CapabilityResourceType;
  operation: CapabilityOperation;
  queryVersion: string;
  normalizedExpression: string;
  includeTerms: string[];
  excludeTerms: string[];
  accountFilters: string[];
  urlInputs: string[];
  limitations: string[];
};

export type RouteRequirementDto = {
  requirement_ref: string;
  scope_keys: string[];
  step_refs: string[];
  platform: CapabilityPlatform;
  resource_type: CapabilityResourceType;
  operation: CapabilityOperation;
  purpose: WorkflowPlannerPurpose;
  regions: string[];
  required_fields: string[];
  optional_fields: string[];
  budget_ceiling: BudgetCeilingDto | null;
  freshness_requirement: string | null;
  rate_limit_requirement: RateLimitIntentDto | null;
  retention_requirement: RetentionIntentDto | null;
  allow_partial_degradation: boolean;
  precondition_failures: DecisionReasonDto[];
};

export type RouteRequirement = {
  requirementRef: string;
  scopeKeys: string[];
  stepRefs: string[];
  platform: CapabilityPlatform;
  resourceType: CapabilityResourceType;
  operation: CapabilityOperation;
  purpose: WorkflowPlannerPurpose;
  regions: string[];
  requiredFields: string[];
  optionalFields: string[];
  budgetCeiling: BudgetCeiling | null;
  freshnessRequirement: string | null;
  rateLimitRequirement: RateLimitIntent | null;
  retentionRequirement: RetentionIntent | null;
  allowPartialDegradation: boolean;
  preconditionFailures: DecisionReason[];
};

export type ScoreBreakdownDto = {
  raw_dimensions: Record<string, number>;
  effective_dimensions: Record<string, number>;
  weights: Record<string, number>;
  weighted_score: number;
  trace_codes: string[];
};

export type ScoreBreakdown = {
  rawDimensions: Record<string, number>;
  effectiveDimensions: Record<string, number>;
  weights: Record<string, number>;
  weightedScore: number;
  traceCodes: string[];
};

export type RouteCandidateDecisionDto = {
  assertion_id: string;
  implementation_id: string;
  capability_status: CapabilityStatus;
  score_breakdown: ScoreBreakdownDto | null;
  weighted_score: number | null;
  route_eligible: boolean;
  readiness_status: WorkflowPlannerAuthReadiness;
  approval_required: boolean;
  approval_reasons: DecisionReasonDto[];
  missing_optional_fields: string[];
  evidence_refs: string[];
};

export type RouteCandidateDecision = {
  assertionId: string;
  implementationId: string;
  capabilityStatus: CapabilityStatus;
  scoreBreakdown: ScoreBreakdown | null;
  weightedScore: number | null;
  routeEligible: boolean;
  readinessStatus: WorkflowPlannerAuthReadiness;
  approvalRequired: boolean;
  approvalReasons: DecisionReason[];
  missingOptionalFields: string[];
  evidenceRefs: string[];
};

export type ShadowRuleDto = {
  enabled: boolean;
  fallback_implementation_id: string | null;
  sample_rate: number | null;
  max_items: number | null;
  reason: string;
  execution_authorized: false;
};

export type ShadowRule = {
  enabled: boolean;
  fallbackImplementationId: string | null;
  sampleRate: number | null;
  maxItems: number | null;
  reason: string;
  executionAuthorized: false;
};

export type RoutePlanPreviewDto = {
  requirement_ref: string;
  status: RoutePlanStatus;
  primary_implementation: RouteCandidateDecisionDto | null;
  fallback_implementations: RouteCandidateDecisionDto[];
  shadow_rule: ShadowRuleDto;
  required_fields: string[];
  optional_fields: string[];
  missing_optional_fields: string[];
  budget_status: WorkflowBudgetStatus;
  rate_limit_policy: RateLimitIntentDto | null;
  retention_policy: RetentionIntentDto | null;
  route_eligible: boolean;
  readiness_status: WorkflowPlannerAuthReadiness | null;
  approval_required: boolean;
  approval_reasons: DecisionReasonDto[];
  policy_gates: DecisionReasonDto[];
  score_breakdown: ScoreBreakdownDto | null;
  exclusion_reasons: DecisionReasonDto[];
  degradation_rule: DecisionReasonDto | null;
  limitations: string[];
  execution_authorized: false;
};

export type RoutePlanPreview = {
  requirementRef: string;
  status: RoutePlanStatus;
  primaryImplementation: RouteCandidateDecision | null;
  fallbackImplementations: RouteCandidateDecision[];
  shadowRule: ShadowRule;
  requiredFields: string[];
  optionalFields: string[];
  missingOptionalFields: string[];
  budgetStatus: WorkflowBudgetStatus;
  rateLimitPolicy: RateLimitIntent | null;
  retentionPolicy: RetentionIntent | null;
  routeEligible: boolean;
  readinessStatus: WorkflowPlannerAuthReadiness | null;
  approvalRequired: boolean;
  approvalReasons: DecisionReason[];
  policyGates: DecisionReason[];
  scoreBreakdown: ScoreBreakdown | null;
  exclusionReasons: DecisionReason[];
  degradationRule: DecisionReason | null;
  limitations: string[];
  executionAuthorized: false;
};

export type StepDataContractFieldDto = {
  name: string;
  data_type: string;
  cardinality: string;
  required: boolean;
  source_step_ref: string | null;
  description: string;
};

export type StepDataContractField = {
  name: string;
  dataType: string;
  cardinality: string;
  required: boolean;
  sourceStepRef: string | null;
  description: string;
};

export type StepDataContractDto = {
  schema_version: string;
  fields: StepDataContractFieldDto[];
};

export type StepDataContract = {
  schemaVersion: string;
  fields: StepDataContractField[];
};

export type WorkflowStepPreviewDto = {
  step_ref: string;
  template_key: string;
  sequence: number;
  label: string;
  execution_kind: "planner_internal" | "future_capability";
  depends_on: string[];
  platform: CapabilityPlatform | null;
  scope_keys: string[];
  resource_type: CapabilityResourceType | null;
  operation: CapabilityOperation | null;
  requirement_ref: string | null;
  input_contract: StepDataContractDto;
  output_contract: StepDataContractDto;
  planning_status: WorkflowStepPlanningStatus;
  limitations: string[];
};

export type WorkflowStepPreview = {
  stepRef: string;
  templateKey: string;
  sequence: number;
  label: string;
  executionKind: "planner_internal" | "future_capability";
  dependsOn: string[];
  platform: CapabilityPlatform | null;
  scopeKeys: string[];
  resourceType: CapabilityResourceType | null;
  operation: CapabilityOperation | null;
  requirementRef: string | null;
  inputContract: StepDataContract;
  outputContract: StepDataContract;
  planningStatus: WorkflowStepPlanningStatus;
  limitations: string[];
};

export type DecisionTraceEntryDto = {
  code: string;
  reason: string;
  scope_keys: string[];
  requirement_ref: string | null;
  details: Record<string, PlannerJsonValue>;
};

export type DecisionTraceEntry = {
  code: string;
  reason: string;
  scopeKeys: string[];
  requirementRef: string | null;
  details: Record<string, PlannerJsonValue>;
};

export type DecisionTraceDto = {
  semantic_entries: DecisionTraceEntryDto[];
  input_diagnostics: DecisionTraceEntryDto[];
};

export type DecisionTrace = {
  semanticEntries: DecisionTraceEntry[];
  inputDiagnostics: DecisionTraceEntry[];
};

export type CoverageSummaryDto = {
  total_requirements: number;
  resolved_requirements: number;
  partial_requirements: number;
  held_requirements: number;
};

export type CoverageSummary = {
  totalRequirements: number;
  resolvedRequirements: number;
  partialRequirements: number;
  heldRequirements: number;
};

export type BudgetSummaryDto = {
  currency: "USD";
  known_selected_unit_cost: string | null;
  unknown_count: number;
  budget_status: WorkflowBudgetStatus;
};

export type BudgetSummary = {
  currency: "USD";
  knownSelectedUnitCost: string | null;
  unknownCount: number;
  budgetStatus: WorkflowBudgetStatus;
};

export type AttributionContractDto = {
  matched_scope_id: string;
  matched_term: string;
  match_reason: string;
  query_version: string;
  requirement_ref: string;
  route_plan_ref: string;
};

export type AttributionContract = {
  matchedScopeId: string;
  matchedTerm: string;
  matchReason: string;
  queryVersion: string;
  requirementRef: string;
  routePlanRef: string;
};

export type WorkflowPlanPreviewDto = {
  schema_version: "workflow_plan_preview.v1";
  planner_contract_version: string;
  project_id: string;
  flow_mode: WorkflowPlannerMode;
  planning_status: WorkflowPlanningStatus;
  normalized_input: NormalizedPlanningInputDto;
  scope_ref_map: ScopeRefMappingDto[];
  query_terms: QueryTermDto[];
  compiled_queries: CompiledPlatformQueryDto[];
  steps: WorkflowStepPreviewDto[];
  route_requirements: RouteRequirementDto[];
  route_plans: RoutePlanPreviewDto[];
  coverage: CoverageSummaryDto;
  budget_summary: BudgetSummaryDto;
  limitations: string[];
  decision_trace: DecisionTraceDto;
  attribution_contract: AttributionContractDto;
  catalog_snapshot_id: string;
  policy_version: string;
  mode_template_version: string;
  query_versions: Partial<Record<CapabilityPlatform, string>>;
  preview_fingerprint: string;
  execution_authorized: false;
  provider_call: false;
  actor_run: false;
  browser_run: false;
  llm_call: false;
  workflow_run_created: false;
  database_write: false;
  generated_at: string;
  request_id: string;
};

export type WorkflowPlanPreview = {
  schemaVersion: "workflow_plan_preview.v1";
  plannerContractVersion: string;
  projectId: string;
  flowMode: WorkflowPlannerMode;
  planningStatus: WorkflowPlanningStatus;
  normalizedInput: NormalizedPlanningInput;
  scopeRefMap: ScopeRefMapping[];
  queryTerms: QueryTerm[];
  compiledQueries: CompiledPlatformQuery[];
  steps: WorkflowStepPreview[];
  routeRequirements: RouteRequirement[];
  routePlans: RoutePlanPreview[];
  coverage: CoverageSummary;
  budgetSummary: BudgetSummary;
  limitations: string[];
  decisionTrace: DecisionTrace;
  attributionContract: AttributionContract;
  catalogSnapshotId: string;
  policyVersion: string;
  modeTemplateVersion: string;
  queryVersions: Partial<Record<CapabilityPlatform, string>>;
  previewFingerprint: string;
  executionAuthorized: false;
  providerCall: false;
  actorRun: false;
  browserRun: false;
  llmCall: false;
  workflowRunCreated: false;
  databaseWrite: false;
  generatedAt: string;
  requestId: string;
};
